### ISSUE TITLE
Cannot add an external blog

### ISSUE THREAD

**Issue by paulrobertlloyd** [microdotblog/sunlit#124]
Originally I thought this issue might relate to not being able to add a second external blog (or at least not being able to replace an existing external blog). However, deleting the app and its data, and trying to set up my first external blog, turns out it’s not possible to add an external blog *at all*!

The authentication process completes, but returning to the app, the settings screen does not update. Quitting the app and restarting it doesn’t fix this issue, either. (Sunlit 3.1, iOS 14.0)

**Comment by cheesemaker** [comment]
Ok, so I think I see what needs to happen. I don't know if it's a problem, but when I'm testing against the Micropub.rocks implementation, the me is coming back as "undefined" so I don't know which Micropub server this token url is for. Example:

sunlit://micropub?code=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ0eXBlIjoiYXV0aG9yaXphdGlvbl9jb2RlIiwic2NvcGUiOiJjcmVhdGUiLCJtZSI6Imh0dHBzOlwvXC9taWNyb3B1Yi5yb2Nrc1wvY2xpZW50XC9EUHJpbDhCNlVKRHgyMjlUIiwiY2xpZW50X2lkIjoiaHR0cHM6XC9cL3N1bmxpdC5pb1wvIiwicmVkaXJlY3RfdXJpIjoiaHR0cHM6XC9cL3N1bmxpdC5pb1wvbWljcm9wdWJcL3JlZGlyZWN0Iiwic3RhdGUiOiI3NTc2OUFBQy1EM0VFLTQ2N0YtQjdGMC1EOTI2MzhFNENBMTAiLCJjcmVhdGVkX2F0IjoxNjAyNDUwODAyLCJleHAiOjE2MDI0NTA4NjIsIm5vbmNlIjoiaWFqOUFSeGIxdyJ9.45FL7B4bRIKoURheb0hjKdtRl-frFXRnTAQIyp_qftA&state=75769AAC-D3EE-467F-B7F0-D92638E4CA10&me=undefined

However, I think if I store the state variable from the initial query and store a mapping to the server URL, I can look it up in reverse. Does that sound reasonable?

**Comment by paulrobertlloyd** [comment]
Hmm, that definitely seems odd, though not sure why! As far as I’m aware, this was all working correctly in 3.0.

**Comment by cheesemaker** [comment]
Ok, I have fixed this in the 3.2 codebase. Not sure if we will do a 3.1.1 or just push forward with 3.2

**Comment by paulrobertlloyd** [comment]
@cheesemaker Awesome, thanks! Will it be in an upcoming Testflight release?

**Comment by cheesemaker** [comment]
Yes, it should be in a 3.2 beta asap. I'm not sure if we will back port it to a 3.1.1 release or just work to get 3.2 out.

**Comment by cheesemaker** [comment]
Is this resolved now? If so, I will close.

**Comment by paulrobertlloyd** [comment]
Sadly not; authentication flow works but external blog doesn’t appear in the settings (even on closing the app).

**Comment by cheesemaker** [comment]
If you delete and reinstall Sunlit from the Beta does it still not show? Just trying to narrow this down as it works fine against Micropub.rocks. Trying to figure out where it's failing/breaking.

**Comment by paulrobertlloyd** [comment]
Trying to delete the app’s data to start again from fresh, but every time I delete the app and reinstall it, I find it in the same state as before. Probably related to #133. Some data persistence issue perhaps?

**Comment by cheesemaker** [comment]
So just to confirm, you've completely deleted Sunlit from your phone (not just the data but the entire app) and reinstalled the beta? And when you do that, you're able to go through authentication but after that it still doesn't show in the list of blogs in settings?

**Comment by paulrobertlloyd** [comment]
Have uninstalled as far as I know… long tap on icon, delete app, and then reinstalled from Testflight.

**Comment by cheesemaker** [comment]
Ok. I'm sort of stumped. What's your server URL? I can hit it and verify that Sunlit is parsing the URL's correctly.

**Comment by cheesemaker** [comment]
Ok, actually, I think the hint to this was in #130 
The Micropub standard removed the `me` param for the final authorization so we updated Sunlit to match, however, my guess is that your implementation hasn't been updated to reflect this. I've added back in `me` so hopefully this will work for you again. Please try the new beta 3.2 (build 381) and see if it works now.

**Comment by cheesemaker** [comment]
@paulrobertlloyd can you try the latest beta and see if your issue is resolved? It fixed it for at least one other user that had reported a similar problem.

**Comment by paulrobertlloyd** [comment]
Apologies for not getting back to this sooner, as wanted to test the original issue that I had before this issue with IndieAuth was revealed, but could only do that once back at my desk. That issue was:

> not being able to add a second external blog (or at least not being able to replace an existing external blog)

I was about to say that this issue is not resolved, but testing adding my two different blogs in a different order reveals that it might be something else. So…

* I can add https://paulrobertlloyd.com (whose Micropub endpoint is https://indiekit.paulrobertlloyd.com)
* I cannot add https://getindiekit.github.io/sandbox/ (whose Micropub endpoint is https://indiekit.herokuapp.com)

The later URL redirects to https://getindiekit.com/sandbox/, yet even if I add that as the URL, it still fails to appear, even though the authentication flow completes. So I’m wondering, might this be an issue with domains that contain a path component?

**Comment by cheesemaker** [comment]
So it looks like it's successfully interrogating https://getindiekit.com/sandbox/
From that we get:
Micropub endpoint - "https://indiekit.herokuapp.com/micropub"
Token endpoint - "https://tokens.indieauth.com/token"
Auth endpoint - "https://indieauth.com/auth"

All of these look correct.

Looking at https://paulrobertlloyd.com
From that we get:
Micropub endpoint - "https://indiekit.paulrobertlloyd.com/micropub"
Token endpoint - "https://tokens.indieauth.com/token"
Auth endpoint - "https://indieauth.com/auth"

All of those appear to be getting saved properly as well. I'm not sure. I can't get past the PGP verification to check the first URL any further.

**Comment by paulrobertlloyd** [comment]
As https://getindiekit.github.io/sandbox/ and https://indiekit.herokuapp.com/micropub are test servers, happy to set up some kind of authentication bypass if that would help? Not sure what’s the best way of doing this however.

**Comment by cheesemaker** [comment]
I'll try anything. I just can't get past the PGP verification using that link on my phone. I have a PGP key setup in my Github account, but I don't understand what that step is even asking me to do.

**Comment by lildude** [comment]
> The later URL redirects to https://getindiekit.com/sandbox/, yet even if I add that as the URL, it still fails to appear, even though the authentication flow completes. So I’m wondering, might this be an issue with domains that contain a path component?

I think you're onto something here. I can't add https://lildude.github.io/dev-micropub-pages/ but I can add https://lildude.co.uk/ - they both use the same token and auth endpoint and no errors are occurring on the auth endpoint side of things.

I can add https://lildude.github.io/dev-micropub-pages/ to the iOS micro.blog app (ignoring the fact it seems to only allow me to add one external blog regardless of the number of accounts).

**Comment by pdxmph** [comment]
I’m having the same issue. Auth works fine, but when I return to the app, the external blog isn’t there. What information can I provide to help troubleshoot?

**Comment by cheesemaker** [comment]
@pdxmph what is the URL of your site?

**Comment by pdxmph** [comment]
> @pdxmph what is the URL of your site?

https://mike.puddingbowl.org

I've gotten this config to work with the micro.blog iOS app (though that complains that it can't find a media endpoint). 

If it helps, I'm using this project to provide the service:

https://github.com/voxpelli/webpage-micropub-to-github

I'm running it out of the box on Heroku, per the README.

**Comment by cheesemaker** [comment]
Hmm. Ok, I'll take a look. I can definitely tell you that if Micro.blog can't find the media endpoint, Sunlit will largely be dead in the water. That being said, Sunlit should have better error handling/reporting in that instance.
